### PR TITLE
Websocket improve retry mechanism /SSS 77

### PR DESCRIPTION
**Pull request by ACC-Vaibhav-Palvi**
Improve the retry mechanism by giving user privilage to add delay and period of their.

### CODE REVIEW

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java [hunk 1]
```diff
@@ -30,7 +30,10 @@
 @Slf4j
 public class SmartStreamTicker {
 
-	private static final int PING_INTERVAL = 10000; // 10 seconds
+	private static int PING_INTERVAL = 10000; // 10 seconds
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: make variable name logically which can show seconds, ms etc.

File: src/main/java/com/angelbroking/smartapi/utils/Utils.java [hunk 2]
```diff
@@ -9,6 +9,10 @@ public static boolean isEmpty(final CharSequence cs) {
 		return cs == null || cs.length() == 0;
 	}
 
+	public static boolean isEmpty(final Number nm) {
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/utils/Utils.java]: why number obj is needed?

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java [hunk 3]
```diff
@@ -65,6 +68,30 @@ public SmartStreamTicker(String clientId, String feedToken, SmartStreamListener
...
+	 *
+	 * @param clientId            - the client ID used for authentication
+	 * @param feedToken           - the feed token used for authentication
+	 * @param delay               - delay in milliseconds
+	 * @param period              - period in milliseconds
+	 * @param smartStreamListener - the SmartStreamListener for receiving callbacks
+	 * @throws IllegalArgumentException - if the clientId, feedToken, or SmartStreamListener is null or empty
+	 */
+	public SmartStreamTicker(String clientId, String feedToken, SmartStreamListener smartStreamListener, Integer delay, Integer period ) {
+		if (Utils.isEmpty(clientId) || Utils.isEmpty(feedToken) || Utils.isEmpty(delay) || Utils.isEmpty(period) ||  Utils.validateInputNullCheck(smartStreamListener)) {
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: use stringutils.isempty not custom function?

**Review by anurag-amx — CHANGES_REQUESTED** [review]

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java (hunk 3 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: using isEmpty() of String class now to check whether the given string is empty or not.

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java [hunk 4]
```diff
@@ -27,7 +26,10 @@
 @Slf4j
 public class SmartStreamTicker {
 
-	private static final int PING_INTERVAL = 10000; // 10 seconds
+	private static int PING_INTERVAL_IN_SECONDS = 10000; // 10 seconds
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: These are milliseconds not seconds. please change variable names.

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java [hunk 5]
```diff
@@ -52,7 +54,7 @@ public class SmartStreamTicker {
      * @throws IllegalArgumentException - if the clientId, feedToken, or SmartStreamListener is null or empty
      */
     public SmartStreamTicker(String clientId, String feedToken, SmartStreamListener smartStreamListener) {
-        if (Utils.isEmpty(clientId) || Utils.isEmpty(feedToken) ||  Utils.validateInputNullCheck(smartStreamListener)) {
+        if (clientId.isEmpty() || feedToken.isEmpty() ||  Utils.validateInputNullCheck(smartStreamListener)) {
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: use Strings.isEmpty(string) this takes of null value as well.

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java [hunk 6]
```diff
@@ -63,12 +65,36 @@ public SmartStreamTicker(String clientId, String feedToken, SmartStreamListener
...
+	 * @param period              - period in milliseconds
+	 * @param smartStreamListener - the SmartStreamListener for receiving callbacks
+	 * @throws IllegalArgumentException - if the clientId, feedToken, or SmartStreamListener is null or empty
+	 */
+	public SmartStreamTicker(String clientId, String feedToken, SmartStreamListener smartStreamListener, Integer delay, Integer period ) {
+		if (clientId.isEmpty() || feedToken.isEmpty() || Utils.isEmpty(delay) || Utils.isEmpty(period) ||  Utils.validateInputNullCheck(smartStreamListener)) {
+			throw new IllegalArgumentException(
+					"clientId, feedToken and SmartStreamListener should not be empty or null");
+		}
+		this.DELAY_IN_SECONDS = delay;
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: if Delay is classs level variable then use small case format.

File: src/main/java/com/angelbroking/smartapi/utils/Utils.java [hunk 7]
```diff
@@ -5,11 +5,11 @@ private Utils() {
...
 	}
 
-	public static boolean isEmpty(final CharSequence cs) {
-		return cs == null || cs.length() == 0;
+	public static boolean isEmpty(final Integer nm) {
+		return nm == null || nm.equals(0);
 	}
 
-	public static boolean isNotEmpty(final CharSequence cs) {
+	public static boolean isNotEmpty(final Integer cs) {
```
Comment (anurag-amx) [inline src/main/java/com/angelbroking/smartapi/utils/Utils.java]: No need for this method. Unnecessary duplicating code.

File: src/test/java/com/angelbroking/smartapi/smartstream/SmartStreamTickerTest.java [hunk 8]
```diff
@@ -61,6 +61,25 @@ void testSmartStreamTicketLTP() throws WebSocketException, InterruptedException
 
 	}
 
+	@Test
+	void testSmartStreamTicket_WithInitialDelayAndPeriod() throws WebSocketException, InterruptedException {
+		try {
+			SmartStreamTicker ticker = new SmartStreamTicker(clientID, feedToken, new SmartStreamListenerImpl(), 6000, 6000);
+			ticker.connect();
+			ticker.subscribe(SmartStreamSubsMode.SNAP_QUOTE, getTokens());
+//			ticker.subscribe(SmartStreamSubsMode.SNAP_QUOTE, getTokens());
```
Comment (anurag-amx) [inline src/test/java/com/angelbroking/smartapi/smartstream/SmartStreamTickerTest.java:71]: Remove unused code

**Review by anurag-amx — CHANGES_REQUESTED** [review]

**Review by gulshan178 — DISMISSED** [review]

File: src/test/java/com/angelbroking/smartapi/smartstream/SmartStreamTickerTest.java (hunk 8 above)
Comment (ACC-Vaibhav-Palvi) [reply src/test/java/com/angelbroking/smartapi/smartstream/SmartStreamTickerTest.java:71]: This code is not unused, it is used in test case where we implement the SmartApiTicker websocket.

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java (hunk 1 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: Variables names are now updated according to its usage. Please check.

File: src/main/java/com/angelbroking/smartapi/utils/Utils.java (hunk 2 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/utils/Utils.java]: Number object is removed now, now it'll onlt check for Integer value object.

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java (hunk 3 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: Now using StringUtils.isEmpty() which will check for both null and empty value for string .

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java (hunk 4 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: Variables names are changed accordingly. Please check.

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java (hunk 5 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: Changes done.

File: src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java (hunk 6 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/smartstream/ticker/SmartStreamTicker.java]: Changes done, now the class level variables are in small cases format only.

File: src/main/java/com/angelbroking/smartapi/utils/Utils.java (hunk 7 above)
Comment (ACC-Vaibhav-Palvi) [reply src/main/java/com/angelbroking/smartapi/utils/Utils.java]: it is removed now.

**Review by anurag-amx — APPROVED** [review]

**Pull request merged**